compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3553,6 +3555,71 @@ TR::Register *OMR::Power::TreeEvaluator::vmulInt32Helper(TR::Node *node, TR::Cod
    return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::vmuluwm);
    }
 
+TR::Register *OMR::Power::TreeEvaluator::vmulInt64Helper(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   if (cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P10))
+      return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::vmulld);
+
+   /*
+   To perform the multiplication without the vmulld instruction, we will break it into parts.
+
+   Suppose we have two long integers (64 bits), AB and CD, where each of A, B, C, and D is one word (32 bits).
+   To get their product, AB * CD, we can do the following:
+
+   AB * CD = B*D + 2^32*(AD + BC) + 2^64*(A*C)
+
+   Since we are only keeping the lower 64 bits of the result, this is equivalent to:
+
+   AB * CD = B*D + 2^32*(AD + BC)
+   */
+
+   TR::Node *firstChild;
+   TR::Node *secondChild;
+   TR::Register *lhsReg, *rhsReg;
+   TR::Register *productReg;
+   TR::Register *tempA, *tempB;
+   TR::Register *shiftReg;
+
+   firstChild = node->getFirstChild();
+   secondChild = node->getSecondChild();
+
+   lhsReg = cg->evaluate(firstChild);
+   rhsReg = cg->evaluate(secondChild);
+
+   productReg = cg->allocateRegister(TR_VRF);
+   tempA = cg->allocateRegister(TR_VRF);
+   tempB = cg->allocateRegister(TR_VRF);
+   shiftReg = cg->allocateRegister(TR_VRF);
+
+   node->setRegister(productReg);
+
+   //set shift amount to 16 bits (will achieve factor of 2^32 by shifting twice)
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::vspltisw, node, shiftReg, -16);
+
+   //productReg = B*D
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vmulouw, node, productReg, lhsReg, rhsReg);
+
+   //tempB = 2^32*(AD + BC)
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vrld, node, rhsReg, rhsReg, shiftReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vrld, node, rhsReg, rhsReg, shiftReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vmulouw, node, tempA, lhsReg, rhsReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vmuleuw, node, tempB, lhsReg, rhsReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vaddudm, node, tempB, tempA, tempB);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vsld, node, tempB, tempB, shiftReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vsld, node, tempB, tempB, shiftReg);
```

Comment:
This won't work. when -16 is splatted into `shiftReg` it gets sign extended to `0xFFFFFFF0`. `vsld` just takes the lower 6 bits as an unsigned value which becomes positive 48. Shifting left by 48 bits twice will always end with `tempB` containing all 0s.

Unfortunately, `vspltisw` does not accept positive 16 as an immediate value. So you'll need to negate the value. I believe `vnegd` will do this for Power9. For Power8 you might need to do something like what `vnegInt32Helper` which is a combination of `xxlxor` and `vsubuwm`.

Reply:
Actually, I thought of something better. Just added the vector of -16 to itself to get a vector of -32. That can then be used to perform shifts or rotates of 32.